### PR TITLE
Add first draft of talk

### PR DESCRIPTION
**Pull request by matthewfeickert**
Add a first real draft of talk for review by @lukasheinrich and @kratsg.

Talk is here: https://matthewfeickert.github.io/talk-SciPy-2020/index.html

### CODE REVIEW

**Comment by kratsg** [comment]
Comments:

- slide 5, add a slide before showing the Higgs boson peak, explain that it was announced with "5-sigma" (quantify that) and then ask the question: "what does that mean that we've found statistical significant evidence?" -- and bring up "counting experiments".... then bridge that into current slide 5 saying: "and we still continue to look for new physics at the LHC with the same statistical techniques"
- slide 6, nobody knows what exotic physics or supersymmetry are
- slide 7, nobody knows what the other parts of histfactory models are -- you might want to start introducing the concept by explaining the two Gaussian example, and explaining how to distinguish them, and lead that up to the general formulation [e.g. nuisance parameters are new for this audience probably]
- slide 8, "until now" -> "until now (2018)"
- slide 9, L(theta|x) is backwards -> Ln(x|\vec{phi}) [note that it's cleaner to write the right hand side as `model.make_pdf(parameters).log_prob(data)`, can you fix code example too?]
- slide 10 should be fixed, indicate that this was made using TensorFlow backend
- slide 11, likelihood function swapped, update code example -- maybe need to add more explanation on to why that's the objective function (-2 ln); never mentioned optimizer before and not showing an example of this in the code example
- slide 12, there's more variables introduced that you didn't explain (`$mu$`, `$\hat{mu}$`, `$\hat{theta}$`, `$\Lambda$`)
- slide 13: mention that there's a JSON schema describing this specification (link to it)

Might want to mention patchsets perhaps (we're updating all hepdata records to patchsets instead of patches).

**Comment by matthewfeickert** [comment]
@kratsg Thanks for taking time to give me feedback so quickly. I know everyone is busy right now, so I really appreciate it! :bow: 

> - slide 5, add a slide before showing the Higgs boson peak, explain that it was announced with "5-sigma" (quantify that) and then ask the question: "what does that mean that we've found statistical significant evidence?" -- and bring up "counting experiments".... then bridge that into current slide 5 saying: "and we still continue to look for new physics at the LHC with the same statistical techniques"

I'm a bit hesitant to take time to introduce our standard of 5-sigma (something that I don't think the audience is familiar with), as it is reasonable to assume that the audience know something about statistics and the concept of statistical significance in general. I can understand that "Search for new physics" with a Higgs plot is maybe not the best, but my plan was to say something along the lines of "Back in 2012 the ATLAS and CMS experiments at the LHC announced discovery of a new particle that was consistent with the properties of the predicted Higgs boson. This discovery of new physics was made possible by the statistical techniques that were employed."

> - slide 6, nobody knows what exotic physics or supersymmetry are

Sure, but what would you put there instead? Would you just show one plot (instead of two) and say "new physics (beyond Standard Model)" beneath it?

> - slide 7, nobody knows what the other parts of histfactory models are -- you might want to start introducing the concept by explaining the two Gaussian example, and explaining how to distinguish them, and lead that up to the general formulation [e.g. nuisance parameters are new for this audience probably]

What is the two Gaussian example? Sorry I think I'm just not remembering what example in particular you mean. I also think that most people in the audience will have performed some form of fit before in their life and so it wouldn't be to strange to think they know what a nuisance parameter is, though I can mention this verbally.

> - slide 8, "until now" -> "until now (2018)"

Done.

> - slide 9, L(theta|x) is backwards -> Ln(x|\vec{phi}) [note that it's cleaner to write the right hand side as `model.make_pdf(parameters).log_prob(data)`, can you fix code example too?]

I don't think it is backwards. The likelihood is a function of the model parameters (with the observed data now being fixed). I do understand suggesting the use of `\vec{\phi}` to match the `pyhf` documentation, but as I'm trying to give the audience a common mathematical object to grasp onto I think it is better to use the much more common `\theta` that is abundant in intro stats literate (c.f. [Wikipedia's likelihood function entry](https://en.wikipedia.org/wiki/Likelihood_function)). Given this I'm not sure if the alternative API call is really cleaner, though I could see the argument if the likelihood was a function of the data with fixed model parameters.

> - slide 10 should be fixed, indicate that this was made using TensorFlow backend

@lukasheinrich made this using [Graphviz](https://www.graphviz.org/) and not [TensorBoard](https://www.tensorflow.org/tensorboard). Does the backend even matter here though?

> - slide 11, likelihood function swapped, update code example -- maybe need to add more explanation on to why that's the objective function (-2 ln); never mentioned optimizer before and not showing an example of this in the code example

Added "(maximizing the likelihood)". I can add a slide on the optimizers, but not sure if that is too important given that people at SciPy probably know that minimizers involve optimizers.

> - slide 12, there's more variables introduced that you didn't explain (`$mu$`, `$\hat{mu}$`, `$\hat{theta}$`, `$\Lambda$`)

Yeah, I didn't see much point in making a legend of them as they really only show up in this one slide and I'm going to address what is on screen and describe the notation, but then make it clear that the important thing is that "the model parameters are profiled out so that we're looking at the impact of the choice of the parameter of interest." &mdash; or something like that.

> - slide 13: mention that there's a JSON schema describing this specification (link to it)

Added

> JSON Schema [describing HistFactory specification](https://scikit-hep.org/pyhf/likelihood.html#bibliography)

to the slide.

> Might want to mention patchsets perhaps (we're updating all hepdata records to patchsets instead of patches).

Yeah, I was thinking about doing this but I figured I'm already going to be close on time and I'm not sure how much it matters to the talk beyond what I have on slide 15.

**Comment by kratsg** [comment]
> I'm a bit hesitant to take time to introduce our standard of 5-sigma (something that I don't think the audience is familiar with), as it is reasonable to assume that the audience know something about statistics and the concept of statistical significance in general. I can understand that "Search for new physics" with a Higgs plot is maybe not the best, but my plan was to say something along the lines of "Back in 2012 the ATLAS and CMS experiments at the LHC announced discovery of a new particle that was consistent with the properties of the predicted Higgs boson. This discovery of new physics was made possible by the statistical techniques that were employed."

I think we should not make assumptions as we want this talk to be approachable for students as well. Perhaps starting with a toy example of rolling dice, and just say something like "this die has a 6 that shows up 1e-6 more of the time than this die which represents the standard model" -- so we need to make a ton of observations and roll the die enough times to be convinced of which "die model" we have. And then we need to determine the likelihood of ... etc. I'm not sure. I'll think more.

> Sure, but what would you put there instead? Would you just show one plot (instead of two) and say "new physics (beyond Standard Model)" beneath it?

Yes.



> What is the two Gaussian example? Sorry I think I'm just not remembering what example in particular you mean. I also think that most people in the audience will have performed some form of fit before in their life and so it wouldn't be to strange to think they know what a nuisance parameter is, though I can mention this verbally.

This is where you have two partially overlapping gaussians and you just need to compute the difference. I remember this nice talk from Max which could be approachable for intro-level: https://indico.cern.ch/event/676637/contributions/2771270/attachments/1550126/2434880/swiatlow.pdf

> I don't think it is backwards. The likelihood is a function of the model parameters (with the observed data now being fixed). I do understand suggesting the use of `\vec{\phi}` to match the `pyhf` documentation, but as I'm trying to give the audience a common mathematical object to grasp onto I think it is better to use the much more common `\theta` that is abundant in intro stats literate (c.f. [Wikipedia's likelihood function entry](https://en.wikipedia.org/wiki/Likelihood_function)). Given this I'm not sure if the alternative API call is really cleaner, though I could see the argument if the likelihood was a function of the data with fixed model parameters.

I think the other call is possibly cleaner. The wikipedia page does state "\mathcal {L}(\theta \mid x) should not be confused with p(\theta \mid x)" which is where my confusion was from. I think our formalism is fine to show the split calls. Although I might tend to agree that something like `likelihood(data).fit(pars)` is much more what we're trying to do (but it is not quite technically possible, I don't think).



> Added "(maximizing the likelihood)". I can add a slide on the optimizers, but not sure if that is too important given that people at SciPy probably know that minimizers involve optimizers.

Not many would be familiar with minuit ;)

**Comment by lukasheinrich** [comment]
Hi @matthewfeickert -- nice slides! See comments below

* slide6: I think binned / unbinned might be a detail in this context (and maybe not clear what that means to non-HEPers). can just go with "language to build statistical models for HEP"
*slide6: probably can stick with one SM, one BSM example (i'd take SUSY)
* slide7: maybe this is too much detail formula wise that's hard to parse.. maybe add a stack plot that shows clearly what "samples" and "channels" are (the SUSY one from before works well.. each bin is a (1-bin) channel  .. the colors are the samples.. I'd try to connect the visual markers in the plot to the elements in the formula and maybe cut back on the text
* slide8: can remove the "basic structure part" and go straight to "this is a mathematical grammar" 
* slide8: maybe link ROOT
* slide 12: maybe relate hypothesis testing to the science "given the data we have did we discover something yes/no" 
* slide 12: might want to remove qtilde=True .. really a detail and w/o it its smaller
* slide 13: I'd maybe highlight the point about "declarative statistical models"
* slide 13: technically the screenshot is not JSON due to the comments... not sure anyone cares tho
* slide 13: scans is parameter scans probably doesn't connect w/ people.. i'd maybe just do "JSON Patch gives ability to easily mutate model. Think: test a new theory"
* slide 13+14: can probably be one slide. I'd remove the legend + just explain orally that the contour is the "main result" (could color in "exluded physics" / "non-exluded physics" ). Would explain that this solved a 20y struggle in HEP to publish stat models
* slide 20: can we not move tb = pyhf.tensorlib out of the loop?, i'd maybe highlight the matching values with a box or sth to drive home the point

**Comment by matthewfeickert** [comment]
Thanks for the great feedback @lukasheinrich! I've addressed all your points below but I have one question that could use some clarification.

> * slide6: I think binned / unbinned might be a detail in this context (and maybe not clear what that means to non-HEPers). can just go with "language to build statistical models for HEP"

Changed to

> to build statistical models in high energy physics (HEP)

in the first bullet point.

> *slide6: probably can stick with one SM, one BSM example (i'd take SUSY)

Done. For future self reference, the three images that we've used over and over in presentations are taken from the following publications
```
.kol-2-8.center[
<br>
.width-100[[![HIGG-2016-25](figures/HIGG-2016-25.png)](https://atlas.web.cern.ch/Atlas/GROUPS/PHYSICS/PAPERS/HIGG-2016-25/)]
<br>
.bold[Standard Model]
]
.kol-3-8.center[
<br>
.width-100[[![SUSY-2016-16](figures/SUSY-2016-16.png)](https://atlas.web.cern.ch/Atlas/GROUPS/PHYSICS/PAPERS/SUSY-2016-16/)]
<br>
<br>
.bold[Supersymmetry]
]
.kol-3-8.center[
<br>
.width-100[[![EXOT-2016-25](figures/EXOT-2016-25.png)](https://atlas.web.cern.ch/Atlas/GROUPS/PHYSICS/PAPERS/EXOT-2016-25/)]
<br>
.bold[Exotic Physics]
]
```

> * slide7: maybe this is too much detail formula wise that's hard to parse.. maybe add a stack plot that shows clearly what "samples" and "channels" are (the SUSY one from before works well.. each bin is a (1-bin) channel  .. the colors are the samples.. I'd try to connect the visual markers in the plot to the elements in the formula and maybe cut back on the text

Done. This really improves the slide! Thanks! I also annotated the SUSY image, but if you and @kratsg think this makes it unhelpfully cluttered (verging on it) I can just use the original.

> * slide8: can remove the "basic structure part" and go straight to "this is a mathematical grammar"

Done

> * slide8: maybe link ROOT

Done.

> * slide 12: maybe relate hypothesis testing to the science "given the data we have did we discover something yes/no"

Done.

> * slide 12: might want to remove qtilde=True .. really a detail and w/o it its smaller

done.

> * slide 13: I'd maybe highlight the point about "declarative statistical models"

Done. First bullet point is now

> Human & machine readable **declarative** statistical models

> * slide 13: technically the screenshot is not JSON due to the comments... not sure anyone cares tho

Yup. Like I did in the CHEP 2019 talk I will verbally note that "of course these highlighted comments that I've added in by hand aren't valid JSON"

> * slide 13: scans is parameter scans probably doesn't connect w/ people.. i'd maybe just do "JSON Patch gives ability to easily mutate model. Think: test a new theory"

Done.

> * slide 13+14: can probably be one slide. I'd remove the legend + just explain orally that the contour is the "main result" (could color in "exluded physics" / "non-exluded physics" ). Would explain that this solved a 20y struggle in HEP to publish stat models

I added a very short slide (current slide 16) to emphasize solving the publishing problem.

Here do you mean (current) slide 15 and 17? When you suggest

> I'd remove the legend

you're saying to manually paint out the expected and observed limit patches for the roundtrip, ROOT, and pyhf on (current) Slide 17?

> * slide 20: can we not move tb = pyhf.tensorlib out of the loop?, i'd maybe highlight the matching values with a box or sth to drive home the point

(Now Slide 21) We can, but given the way that `pyhf.set_backend` works any existing objects that are `pyhf` backends aren't updated, so this means that `pyhf.tensorlib` needs to get be used _everywhere_. I've changed it, but if you find this worse I can change it back no problem (thank you Git :slightly_smiling_face:).

```python
import pyhf
pyhf.tensorlib.name  # default backend/optimizer is numpy/scipy
means = [5, 8]
stds = [1, 0.5]
values = [4, 9]
for backend in ["numpy", "tensorflow", "pytorch", "jax"]:
    pyhf.set_backend(backend)
    normals = pyhf.tensorlib.normal_dist(
        pyhf.tensorlib.astensor(means), pyhf.tensorlib.astensor(stds)
    )
    normals.log_prob(pyhf.tensorlib.astensor(values))
```
